### DISCUSSION TITLE
[RFC][Draft] Checkout/Order additional fees

### DISCUSSION THREAD

**Discussion by krzysztofwolski** [saleor/saleor#14470] [RFC]
## Problem

TLDR: Add additional fees to the checkout/order.

This issue is a compilation of already reported issues and known workarounds:

- processing fee for using a particular payment method: 
  - https://github.com/saleor/saleor/issues/7526
  - https://github.com/saleor/saleor/issues/10114
- additional delivery fee tax
  - Colorado Retail Delivery Fee in AvaTax https://docs.saleor.io/docs/3.x/developer/app-store/apps/taxes/tax-codes

The mentioned examples are related to different parts of the API (shipping and payments), but I see a common pattern of "Based on condition X, I would like to add an additional fee to the order". Such a use case is a common case for b2b.

Current workaround: 
- Create a digital product hidden from the listings
- Add it to the checkout via webhooks
- Add custom logic in all presentation views to treat it not as "normal" product, but a fee (for example - don't allow to remove it in the cart view)

Problems with the workaround:
- need custom code for synchronization of the fees
- possible impact on reporting and stats - "fee product" has to be removed from calculating best-selling products etc.

## General Assumptions

- if fees could be kept as order lines, they should be easy to filter out (unlike current solution with checking the product type, etc.).
- it might be required to give an option of assigning custom tax classes to the fees, to avoid double taxation 
- not strictly required, but would massively benefit DX/data clarity: there should be a way to mark which lines are related to the additional fee. Usecases:
  - One of the items in the cart require heavy duty shipping. I would like to have a way to display in the checkout view which item is reason for that.

### Acceptance Criteria

- [ ] Customers should not be able to modify it via API
- [ ] There should be an API allowing apps to apply fees
- [ ] (TBD) Multiple fees can be applied
- [ ] Fee requires tax code information

### To be considered, open questions

- How fees should be handled in the dashboard (refunds UI)
- Should the fee be an extension of the normal Order line, or separate array?
- Important edgecase for payment fees: do not forget about split payments! Consider scenario: half of the order total is paid with cash (no fees), and half with credit card (5% fee). It means, that API should not always assume that order total is base for the calculations

**Comment by krzysztofwolski** [comment]
Update: 
- added case for split payments
- added request for marking which line is related to the additional fee
- added case for the taxes
